### ISSUE TITLE
prop to control value and operator reset functionality on field change.

### ISSUE THREAD

**Issue by lakk1** [react-querybuilder/react-querybuilder#121]
Hi,
Using query builder when I update **_operator_** and **_value_**  before _**field**_.
Setting _**field**_  resets  **_operator_** and **_value_** to defaults.
I need to prevent this from happening using a prop.

We have a use-case where _**field**_, _**operator**_ and _**values**_ are independent sets and the current reset mechanism is not expected.

**Comment by jakeboone02** [comment]
You can accomplish what you want with the current functionality, using a combination of custom components, `useEffect`, and the `usePrevious` hook found [here](https://usehooks.com/usePrevious/).

Basically you just ignore the current value and reset it to the previous value when the field changes. Do this for both the `ValueSelector` and the `OperatorSelector` components.

Here's some example code, based off the default create-react-app setup:

```jsx
// usePrevious.js

import { useRef, useEffect } from 'react';

const usePrevious = value => {
  const ref = useRef(null);

  useEffect(() => {
    ref.current = value;
  }, [value]);

  return ref.current;
};

export default usePrevious;
```

```jsx
// ValueEditor.js

import React, { useEffect } from 'react';
import usePrevious from './usePrevious';

const ValueEditor = props => {
  const prevField = usePrevious(props.field);
  const prevValue = usePrevious(props.value);

  useEffect(() => {
    if (prevField !== props.field) {
      props.handleOnChange(prevValue);
    }
  });

  return (
    <input
      value={props.value}
      onChange={ev => props.handleOnChange(ev.target.value)}
    />
  );
};

export default ValueEditor;
```

```jsx
// OperatorSelector.js

import React, { useEffect } from 'react';
import usePrevious from './usePrevious';

const OperatorSelector = props => {
  const prevField = usePrevious(props.field);
  const prevValue = usePrevious(props.value);

  useEffect(() => {
    if (prevField !== props.field) {
      props.handleOnChange(prevValue);
    }
  });

  const options = props.options.map(opt => (
    <option value={opt.name}>{opt.label}</option>
  ));

  return (
    <select
      value={props.value}
      onChange={ev => props.handleOnChange(ev.target.value)}
    >
      {options}
    </select>
  );
};

export default OperatorSelector;
```

```jsx
// App.js

import React from 'react';
import QueryBuilder from 'react-querybuilder';
import OperatorSelector from './OperatorSelector';
import ValueEditor from './ValueEditor';

const App = () => {
  const fields = [
    { name: 'FIELD1', label: 'Field 1' },
    { name: 'FIELD2', label: 'Field 2' },
  ];
  const controlElements = {
    valueEditor: ValueEditor,
    operatorSelector: OperatorSelector,
  };
  const initialQuery = {
    id: 'group',
    combinator: 'AND',
    rules: [
      { id: 'rule1', field: 'FIELD1', operator: '=', value: 'Test 1' },
      { id: 'rule2', field: 'FIELD2', operator: '=', value: 'Test 2' },
    ],
  };

  return (
    <QueryBuilder
      fields={fields}
      onQueryChange={q => console.log(q)}
      controlElements={controlElements}
      query={initialQuery}
    />
  );
};

export default App;
```

If you run that, try to change the field selection and you'll see that the operator and value do not change automatically.

You can get pretty sophisticated with your `useEffect` code here. Where I work, we have a query builder that uses a completely different `ValueEditor` depending on the field and operator selections, so sometimes the value needs to be reset but sometimes it doesn't.

Hope that helps.

**Comment by lakk1** [comment]
@jakeboone02 thank you for the reply we have already implemented the approach you have given above.
But we think it would be good to have first-class support to control reset functionality with the additional property.
May you please look into PR #122.

**Comment by jakeboone02** [comment]
@lakk1 - I have implemented your change in [version 3.1.0](https://www.npmjs.com/package/react-querybuilder).
